File: MasterCatalog.Api/Services/CompanyCatalogService.cs
using MasterCatalog.Dal.Contracts;
using MasterCatalog.Domain.Models;

namespace MasterCatalog.Items.Api.Services
{
    public class CompanyCatalogService : ICompanyCatalogService
    {
        private readonly IItemService _itemService;
        private readonly ICompanyCatalogRepository _companyCatalogRepository;
        private readonly ICompanyRepository _companyRepository;

        public CompanyCatalogService(IItemService itemRepository, ICompanyCatalogRepository companyCatalogRepository, ICompanyRepository companyRepository)
        {
            _itemService = itemRepository;
            _companyCatalogRepository = companyCatalogRepository;
            _companyRepository = companyRepository;
        }

        public CompanyCatalog GetByCompanyIDAndItemID(int companyID, int itemID)
        {
            return _companyCatalogRepository.GetByCompanyIDAndItemID(companyID, itemID);
        }

        public List<CompanyCatalog> GetByItemID(int itemID)
        {
            return _companyCatalogRepository.GetByItemID(itemID);
        }

        public List<CatalogItem> GetCatalogItems(string companyCode)
        {
            var company = _companyRepository.GetAll().FirstOrDefault(x => x.CompanyCode == companyCode);
            if (company == null)
            {
                return new List<CatalogItem>();
            }

            return GetCatalogItems(company.CompanyID.Value);
        }

        public List<CatalogItem> GetCatalogItems(int companyID)
        {
            var catalogItems = new List<CatalogItem>();

            var catalog = _companyCatalogRepository.GetByCompanyID(companyID);
            if (!catalog.Any())
            {
                return catalogItems;
            }

            var items = _itemService.GetItemsGroupByItemID();

            foreach (var catalogEntry in catalog)
            {
                if (!items.ContainsKey(catalogEntry.ItemID))
                {
                    continue;
                }

                var item = items[catalogEntry.ItemID];

                var catalogItem = new CatalogItem()
                {
                    ItemID = catalogEntry.ItemID,
                    CompanyCatalogID = catalogEntry.CompanyCatalogID.Value,
                    CompanyID = companyID,
                    ItemDescription = item.ItemDescription,
                    ItemName = item.ItemName,
                    UnitQuantity = item.UnitQuantity,
                    Upc = item.Upc,
                    RetailPrice = catalogEntry.RetailPrice,
                    UnitOfMeasureID = item.UnitOfMeasureID,
                    UnitOfMeasure = item.UnitOfMeasure,
                    Attributes = item.Attributes,
                    Images = item.Images,
                    Category = item.Category,
                    CategoryID = item.CategoryID
                };

                catalogItems.Add(catalogItem);
            }

            return catalogItems;
        }

        public CompanyCatalog InsertCatalogItem(CompanyCatalog catalogItem)
        {
            _companyCatalogRepository.Insert(catalogItem); 
            return catalogItem;
        }

        public CompanyCatalog UpdateCatalogItem(CompanyCatalog catalogItem)
        {
            _companyCatalogRepository.Update(catalogItem);
            return catalogItem;
        }

        public void DeleteCatalogItem(int companyCatalogID)
        {
            _companyCatalogRepository.Delete(companyCatalogID);
        }
    }
}

File: MasterCatalog.Web/wwwroot/js/services/companyCatalogService.js
import { ApiUtility } from '../utilities/apiHelper.js';
import { CatalogItem } from '../models/catalogItem.js';
import { ItemService } from './itemService.js';

const _api = new ApiUtility('companycatalog');
const _itemService = new ItemService();

export class CompanyCatalogService {

    async insert(obj) {
        const response = await _api.insert(obj);
        return this.mapToModel(response.data);
    }

    async update(obj) {
        const response = await _api.update(obj);
        return this.mapToModel(response.data);
    }

    async delete(id) {
        return _api.delete(id);
    }

    async getByCompany(id) {
        const response = await _api.get({ id });
        const items = response.data.map((obj) => {
            return this.mapToModel(obj);
        });

        return items;
    }

    async getByCompanyAndItem(companyId, itemId) {
        const response = await _api.get({ companyId, itemId });
        return this.mapToModel(response.data);
    }

    async getByItem(itemId) {
        const response = await _api.get(null, 'ByItem/' + itemId);
        const items = response.data.map((obj) => {
            return this.mapToModel(obj);
        });

        return items;
    }

    async items(id) {
        const response = await _api.get(null, 'items/' + id);
        const items = response.data.map((obj) => {
            return this.mapToModel(obj);
        });

        return items;
    }

    mapToModel(obj) {
        const catalogItem = new CatalogItem(
            obj.companyCatalogID,
            obj.companyID,
            obj.itemID,
            obj.retailPrice
        );
        
        const item = _itemService.mapToModel(obj);
        catalogItem.itemName = item.itemName;
        catalogItem.itemDescription = item.itemDescription;
        catalogItem.unitQuantity = item.unitQuantity;
        catalogItem.unitOfMeasureID = item.unitOfMeasureID;
        catalogItem.unitOfMeasure = item.unitOfMeasure;
        catalogItem.images = item.images;
        catalogItem.upc = item.upc;
        catalogItem.categoryID = item.categoryID;
        catalogItem.category = item.category;
        catalogItem.attributes = item.attributes;
        
        return catalogItem;
    }
}